PolyEngine/Core/Src/Collections/BitMask.cpp

```diff
@@ -0,0 +1,382 @@
+		return temp;
+	}
+	//rhs size bigger than lhs size
+	else if (BitList.GetSize() < rhs.BitList.GetSize())
+	{
+		BitMask temp(rhs.BitsNumber);
+		for (size_t i = 0; i < BitList.GetSize(); i++)
+			temp.BitList[i] = BitList[i] ^ rhs.BitList[i];
+		for (size_t i = BitList.GetSize(); i < rhs.BitList.GetSize(); i++)
+			temp.BitList[i] = rhs.BitList[i];
+		return temp;
+	}
+	return BitMask(0);
+}
+
+BitMask BitMask::operator&(const BitMask& rhs) const
+{
+	if (BitList.GetSize() == rhs.BitList.GetSize())
+	{
+		size_t tempBitsNumber = 0;
+		if (BitsNumber > rhs.BitsNumber)
+			tempBitsNumber = BitsNumber;
+		else
+			tempBitsNumber = rhs.BitsNumber;
+
+		BitMask temp(tempBitsNumber);
+		for (size_t i = 0; i < BitList.GetSize(); i++)
+			temp.BitList[i] = BitList[i] & rhs.BitList[i];
+		return temp;
+	}
+	else if (BitList.GetSize() > rhs.BitList.GetSize())
+	{
+		BitMask temp(BitsNumber);
+		for (size_t i = 0; i < rhs.BitList.GetSize(); i++)
+			temp.BitList[i] = BitList[i] & rhs.BitList[i];
+		for (size_t i = rhs.BitList.GetSize(); i < BitList.GetSize(); i++)
+			temp.BitList[i] = ZERO;
+		return temp;
+	}
+	else if (BitList.GetSize() < rhs.BitList.GetSize())
+	{
+		BitMask temp(rhs.BitsNumber);
+		for (size_t i = 0; i < BitList.GetSize(); i++)
+			temp.BitList[i] = BitList[i] & rhs.BitList[i];
+		for (size_t i = BitList.GetSize(); i < rhs.BitList.GetSize(); i++)
+			temp.BitList[i] = ZERO;
+		return temp;
+	}
+	return BitMask(0);
+}
+
+BitMask& BitMask::operator~()
+{
+	for (auto& x : BitList)
+		x = ~x;
+
+	return *this;
+}
+
+
+bool BitMask::Resize(const int offset)
+{
+	if (offset > 0)
+	{
+		if (BitsNumber + offset <= GetDynarraySize()*TYPE_BIT)
+		{
+			BitsNumber += offset;
+			return true;
+		}
+		else
+		{
+			size_t currentSize = BitList.GetSize();
+			size_t targetSize = 0;
+			if ((BitsNumber + offset) % TYPE_BIT)
+				targetSize = (BitsNumber + offset) / TYPE_BIT + 1;
+			else
+				targetSize = (BitsNumber + offset) / TYPE_BIT;
+
+			size_t pushBackCount = targetSize - currentSize;
+
+			for (size_t i = 0; i < pushBackCount; i++)
+				BitList.PushBack(ZERO);
+
+			BitsNumber += offset;
+			return true;
+		}
+	}
+
+	if (offset < 0)
+	{
+		HEAVY_ASSERTE(BitsNumber + offset<BitsNumber, "Out of bounds");
+		if (BitsNumber + offset >(GetDynarraySize() - 1)*TYPE_BIT)
+		{
+			BitsNumber += offset;
+			return true;
+		}
+		else
+		{
+			size_t currentSize = BitList.GetSize();
+			size_t targetSize = 0;
+			if (-1 * (BitsNumber + offset) % TYPE_BIT)
+				targetSize = (BitsNumber + offset) / TYPE_BIT + 1;
+			else
+				targetSize = (BitsNumber + offset) / TYPE_BIT;
+
+			size_t popBackCount = currentSize - targetSize;
+			for (size_t i = 0; i < popBackCount; i++)
+				BitList.PopBack();
+
+			BitsNumber += offset;
+			return true;
+		}
+	}
+
+
+	if (offset == 0)
+	{
+		return false;
+	}
+	return false;
+}
+
+size_t BitMask::BitListIndex(size_t index)
+{
+	return index / TYPE_BIT;
+}
+
+BitMask& BitMask::operator|=(const BitMask& rhs)
```

Comment:
No need to write this like that, simply write `{ *this = *this | rhs; }`. Repeat this for other operators overloads.